### PR TITLE
Improve and enable writing test wrappers to disk

### PR DESCRIPTION
**Pull request by markples**
`CodeBuilder` utility (mainly handles indentation) was lifted from the [PROSE codebase](https://msdata.visualstudio.com/PROSE/_git/PROSE?version=GCa244bb231dcaabe04b4825d4e093beaceed434d8&path=/Microsoft.ProgramSynthesis/Common/Utils/CodeBuilder.cs).  I omitted the last two revisions to it as they opened a few API questions for me and I didn't need the functionality anyway.

Then use it to format the generated code in `XUnitWrapperGenerator`.  Diffs are probably best viewed with whitespace ignored.  Writing simple text (`Append`/`AppendLine`) is the same as `StringBuilder`.  Indentation is specified (usually with a `using` that cleans up) and then added automatically.

Interesting tidbits:
- Factored `AppendAliasMap` from the 3 identical uses
- The loop structure for breaking the tests into groups of 50 doesn't allow `using`.  I did the push/pop of indentation manually rather than trying to restructure it.

Add a new MergedTestRunner.targets for relevant logic.  Enable `EmitCompilerGeneratedFiles` in it.

Questions:
- Should we leave `EmitCompilerGeneratedFiles` off and instead recommend turning it on at the point when someone hits problems and needs to see it?  [agreement to automatically turn it on for merged groups]
- I didn't convert `GenerateXHarnessTestRunner` because I haven't seen that file while working on merged test groups.  Should I?   [Edit: I'll need to convert it or add some CodeBuilder->string transitions to fix things]  [converted]
- Is there something like `CodeBuilder` already in the repo?  It wouldn't be very hard to switch to something else similar and avoid duplication.

### CODE REVIEW

File: src/tests/Common/XUnitWrapperGenerator/CodeBuilder.cs
Comment (trylek) [inline src/tests/Common/XUnitWrapperGenerator/CodeBuilder.cs:1]: This should have the standard copyright header.

File: src/tests/JIT/HardwareIntrinsics/HardwareIntrinsics_r.csproj
Comment (trylek) [inline src/tests/JIT/HardwareIntrinsics/HardwareIntrinsics_r.csproj:4]: I think it's reasonable to emit compiler-generated files for the merged wrappers (as opposed to the simple wrappers where I'd be worried about tons of files being generated potentially degrading Roslyn perf) but I think it's somewhat ugly to need to explicitly put the property on each merged wrapper; could we instead tweak the scripts so that <code>EmitCompilerGeneratedFiles</code> defaults to <code>true</code> when <code>IsMergedTestRunnerAssembly</code> is set to true?

**Review by trylek — APPROVED** [review]
Overall looks great to me, thanks Mark! That indeed makes both the emitter and the emitted code much easier to reason about. I would recommend waiting for Jeremy's response before merging in to make sure I haven't overlooked some aspect he's more familiar with. For the XHarness runner, AFAIK that is only used for Mono testing, I guess that's why you didn't hit that when working on this change.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by markples** [comment]
@trylek @jkoritzinsky Thank you both.  I've run into a (simple?) problem with not setting EmitCompilerGeneratedFiles in each project file.  We want it to be set if IsMergedTestRunnerAssembly is set.  Directory.Build.props is very early - before a test's csproj has set IsMergedTestRunnerAssembly.  However, Directory.Build.targets is very late - after Microsoft.Managed.Core.targets.  In it, EmitCompilerGeneratedFiles is set to false if unspecified, and CompilerGeneratedFilesOutputPath is set iff EmitCompilerGeneratedFiles is set.  We can successfully override EmitCompilerGeneratedFiles in Directory.Build.targets, but then the path in unset which leads to a warning and no output.

I feel like I'm missing something about best practices and conditional defaults.  Is there a typical solution to this?

The best that I can think of is having all proj (or just the ones for merged tests at first?) manually import a new file.  It could
- Set a default for EmitCompilerGeneratedFiles
- Do the item munging that is needed for "extern alias"es to work for the merged wrappers to refer to individual tests
- Replace the line moving MergedWrapperProjectReference to ProjectReference in each proj file
- Consider defaults for "normal" proj (say, include **/*.cs or something like that if nothing has been included already?  I'm less convinced by this one, though in doing the proj rewriting for merged groups, seeing the line adding the same $(whatever).cs into the compile item in almost every proj already had me thinking about a possible default a bit)

**Comment by markples** [comment]
After poking around, `Directory.Build.targets` does appear to indeed to be too late in the imports from the SDK.  I added a new file (src/tests/MergedTestRunner.targets - would welcome new naming.  I chose .targets over .props because it is imported at the end of the .csproj, which seems to be the general convention, though it does not introduce any actual Targets) and put the merged test duplication in there.  I did not try to do something more general (like the 4th bullet in my previous comment).

**Comment by trylek** [comment]
Thanks Mark, I really like this approach! I believe it also opens the door for various future merged wrapper modifications without fighting the overcomplicated recursive build script system every single time.

**Review by jkoritzinsky — APPROVED** [review]

**Review by trylek — APPROVED** [review]

**Comment by markples** [comment]
All failures are the known "NuGet-Migrations" one

**Pull request merged**
